lib/kitchen_sink/algorithms.ex

```diff
@@ -87,31 +63,67 @@ defmodule KitchenSink.Algorithms do
       ...>   end
       ...> end
       iex>
-      iex> {:ok, result} = Algorithms.binary_interval_search(1, 100, solve, 0.0)
+      iex> {:ok, result} = Algorithms.binary_search(1, 100, solve, 0.0, :interval)
       iex> Float.round(result, 6)
       10.311285
   """
-
-  @spec binary_interval_search(
-    binary_interval_search_position, binary_interval_search_position, binary_interval_search_fit_func, any
-  ) :: binary_interval_search_result
-  def binary_interval_search(interval_start, interval_finish, fit, target \\ nil) when is_number(interval_start) and is_number(interval_finish) and is_function(fit) do
-    do_binary_interval_search(interval_start / 1, interval_finish / 1, fit, target)
+  @spec binary_search(
+    binary_search_position, binary_search_position, binary_search_fit_func, any, binary_search_strategy
+  ) :: binary_search_result
+  def binary_search(range_start, range_finish, fit, target, strategy) when is_function(fit) do
+    do_binary_search(range_start, range_finish, fit, target, strategy)
   end
 
-  defp do_binary_interval_search(position, position, fit, target), do: ok_or_not_found(position, fit, target)
-  defp do_binary_interval_search(start, finish, fit, target) when start > finish do
-    binary_interval_search(finish, start, fit, target)
+  defp do_binary_search(start, finish, fit, target, strategy) when finish < start do
+    binary_search(finish, start, fit, target, strategy)
+  end
+  defp do_binary_search(position, position, fit, target, _), do: ok_or_not_found(position, fit, target)
+  defp do_binary_search(start, finish, fit, target, :midpoint) do
+    mid = binary_search_midpoint(start, finish)
+    case fit.(mid, target) do
+      :ok -> {:ok, mid}
+      :high -> do_binary_search(bounded_increment(mid, finish), finish, fit, target, :midpoint)
+      :low -> do_binary_search(start, bounded_decrement(mid, start), fit, target, :midpoint)
+   end
   end
-  defp do_binary_interval_search(start, finish, fit, target) do
-    mid = start + (finish - start) / 2.0
+  defp do_binary_search(start, finish, fit, target, :interval) do
+    mid = binary_search_interval(start, finish)
     case fit.(mid, target) do
       :ok -> {:ok, mid}
-      :high -> do_binary_interval_search(mid, finish, fit, target)
-      :low -> do_binary_interval_search(start, mid, fit, target)
+      :high -> do_binary_search(mid, finish, fit, target, :interval)
+      :low -> do_binary_search(start, mid, fit, target, :interval)
     end
   end
 
+  # `:midpoint` default
+  def binary_search(range_start, range_finish, fit, target) do
+    binary_search(range_start, range_finish, fit, target, :midpoint)
+  end
+
+  # Strategies
+  defp binary_search_midpoint(start, finish) do
+    start + div(finish - start, 2)
+  end
+
+  defp binary_search_interval(start, finish) do
+    start + (finish - start) / 2.0
+  end
+
+  # Increment bounds
+  defp bounded_increment(to_increment, bound) when to_increment >= bound do
```

Comment:
Opinions sought... For "trivial" functions I have seen and liked the style

    defp bounded_increment(to_increment, bound) when to_increment >= bound, do: bound
    defp bounded_increment(to_increment, _), do: to_increment + 1

Reply:
To my eye, this pushes the interesting part of the function out to the right, and makes it harder to vertically scan and understand the difference between these functions. Team, other opinions?

Reply:
I am with @mikestok on this. I find it allows the function definitions to be scanned more easily. Assuming it doesn't push too far out of the width bounds of course.